### ISSUE TITLE
Schedule not working

### ISSUE THREAD

**Issue by tifoji** [emilsvennesson/kodi-fsgo#3]
Opening this one to track schedule as a separate issue, but more importantly I wanted to say that I can help with captures/fiddler stuff if you need it.

Also any interest in parsing channel/all ? (media-api.foxsportsgo.com/epg/ws/channel/all)

**Comment by emilsvennesson** [comment]
Actually, the code for retrieving the schedule is in place already. Thanks for the offer though. :-)

I could use some input on what the best way to present the schedule is. By date would be preferable, but I haven't found a good way of obtaining all upcoming dates. I could of course hard-code it, but that's something I'd like to avoid. I could also grab them by looping through all upcoming events, but I'm unsure about the performance seeing as there can be over 300 events to loop through.

Thoughts?

Regarding the channel URL: Do you know what it's being used for in the app? There's very little info to grab in the JSON response.

**Comment by tifoji** [comment]
There is a very good use of the channel URL I can think of but perhaps I will PM to you on the forums about it.

Regarding the upcoming events, 300+ events seems normal to me and I don't think there would be any performance impact. Let me see if there is a way to limit it by passing an argument to the request.

**Comment by tifoji** [comment]
Well that didn't take long. So in the official app, they are displaying 7 days worth of events (including "today"). The call is 

GET /epg/ws/featured/all/offset/0/size/6 HTTP/1.1

Offset = 0 should be current day 
size = 6 (how many days we want if we want to limit)

Should be easy ;)

**Comment by tifoji** [comment]
Another method to truncate the results

GET media-api.foxsportsgo.com/epg/ws/schedule?start_date=2016-10-09T05%3A00%3A00.0000000Z&end_date=2016-10-16T04%3A59%3A59.0000000Z HTTP/1.1

default seems to be 7 days so changing end_date to something custom like 2-3 days (or ask the user via options etc)

**Comment by emilsvennesson** [comment]
@siuside The first URL you posted is for the 'Featured' section and not the entire schedule. The second URL and its parameters is what I've implemented. I noticed you can send an empty end_date (example: https://media-api.foxsportsgo.com/epg/ws/schedule/?start_date=2016-10-09T05%3A00%3A00.0000000Z&end_date=None) to get the entire upcoming schedule.

 I'll do some tests on this tomorrow. It's time to enjoy some NFL action now. :-)

**Comment by tifoji** [comment]
I ran the request with end_date=None and noticed that it was just providing one extra day's worth of results.

Default end date in app (upto 7 days ahead) ; Content-Length: 1876871 , "total_results":827
EndDate = None (upto 8 days ahead):  Content-Length: 2063419 , "total_results":909

Also the total_results is not really the actual number of events obviously. Still I think we make end_date an option with a max of 7 to make it similar to the app. 

Now to display. I think we can parse through the entire json response and first select airing_date as the base and addDirectory for these ( so at most 7 directories which are nothing but rolling dates). Then when inside a particular date/directory, grab the "title=" of the event , then assign the channe_name/channel_tag to the event title and loop through the entire json. 

So in schedule mode, it would look a bit different from Live where the loop is based on channel_tag.

So once a user clicks schedule in addon, these are the directories that get displayed in next screen.

Today
10-10-2016
10-11-2016
10-12-2016
10-13-2016
10-14-2016
10-15-2016

Now once we click on a particular day, example "Today", the display should be 

title    ,    airing_date  ,   channel_name/channel_tag
or
airing_date , channel_name/channel_tag , title 

etc

Thoughts?

**Comment by emilsvennesson** [comment]
Sounds good. That's what I had in mind as well. :-) Any particular reason why you think we should display it differently than how it currently is in Live though? I think we should be consistent and use the same naming scheme across all sections.

The problematic thing is that we can't really rely on the dates provided in the JSON response (that also applies to start_date and end_date). Those are in UTC and since we localize the time, the date might not be correct depending on the user's timezone.

**Comment by tifoji** [comment]
Nah I was overthinking it a bit with multiple results possible for the same channel_name. Being consistent is a better idea so we could display it the same way. 

Yup I noticed the UTC and figured it would be a bit of a challenge at the 24 hr cutover. I know you will come up with something good anyways and am looking forward to it :+1:

**Comment by emilsvennesson** [comment]
Implemented with 48aa801a858f95fa81d4fc95ae96a88d5d1a273b. I had to do a bit more hacking than I would have liked, but it turned out pretty good I think. 

Any thoughts/feedback before I close this?

**Comment by tifoji** [comment]
I gave it a quick go before heading to work, so not thoroughly tested. One thing I noticed real quick was that last events getting displayed for all days was 6:30 PM Central (my local timezone). So we have a good 6 hours of schedule per day not showing up (time diff between CST & UTC). I am guessing it is affecting other time zone similarly. I can let you know (later today) the last event of the each day if you want to get a better idea.

**Comment by emilsvennesson** [comment]
@siuside Huh, that's weird. Could you try changing line 301 in fslib to `end_date_obj = filter_date_obj + timedelta(days=2)` and see if that fixes it?

**Comment by tifoji** [comment]
That does fix it ! Nice work

**Comment by emilsvennesson** [comment]
I have done some more work on this and reworked the main menu a bit. The 'Featured' section has also been added.

**Comment by tifoji** [comment]
Great work. The layout looks amazing now. We can close this and I think this could very well be a good 1st full release if you wanted to create your repo.

Thanks again :+1:

**Comment by emilsvennesson** [comment]
Yeah, I will probably submit it to the official repository once #1 is done. :-)

**Comment by emilsvennesson** [comment]
@siuside Could you try e48162cfe0f1d195d86dff1311edd2684e9e64fd and see if that fixed the bug you reported in the Kodi thread?

**Comment by tifoji** [comment]
Sure. Looks like one several commits after e48162cfe0f1d195d86dff1311edd2684e9e64fd did you want me to try with latest master instead? I will be able to confirm only after 9 PM CT as that's when the bug hits

**Comment by emilsvennesson** [comment]
@siuside The time just passed midnight here and I just encountered the same bug. It should **now** (!) be fixed with edb1ed9dd8b5697cad0142bccbac463f9b2a2bb4. I am almost certain that all bugs related to the schedule has been fixed but I will leave this open until you've confirmed it as well.

Yeah, you should try with the latest master. Let's just hope the changes I've made today didn't introduce any new bugs...

**Comment by tifoji** [comment]
Sounds good. I will update to latest master and confirm later. Thanks !

**Comment by tifoji** [comment]
Confirmed fixed with latest master https://github.com/emilsvennesson/kodi-fsgo/commit/edb1ed9dd8b5697cad0142bccbac463f9b2a2bb4

Please don't close it yet though. I think there may be a minor bug at the moment still affecting only "Today" where events in the past (i think upto 4 hours or so) are also getting displayed even though a more current program for a particular channel is the current streaming event. Not as bad as the previous bug where you couldn't even get into those channels. This one is more cosmetic

**Comment by tifoji** [comment]
Another timeline I wanted to test was display of events at exactly 12 AM Central Standard Time / 24 hour cutover and it looks great. Events that are active in the prior date close to the 12 AM mark as well as new events from the new date all get displayed successfully using the "Today" option :+1:

**Comment by emilsvennesson** [comment]
@siuside The possible bug you mentioned; were the past events live (listed in green)? Did you also verify with the website/app and compared with how it looked there? I was not able to reproduce this yesterday.

**Comment by tifoji** [comment]
Yes the past events are listed in Green. I will check this again a few hours from now

**Comment by tifoji** [comment]
Yup it is all events in Green. I guess these are still active channel regardless of what the current time is. 

So looks like all bugs squashed :+1:

**Comment by emilsvennesson** [comment]
@siuside Great! It's getting time to cut the first release. :-)
